### PR TITLE
Feature/docker aware dap mode

### PR DESCRIPTION
**Pull request by factyy**
Add support for docker-embedded debuggers (both in containers and images).
Mainly consists of an additional paths translation layer (look into `dap-mode.el` integration) and a config parsing / handling boilerplate code (mostly in `dap-docker.el`).

I don't know however if I have added all the dependencies to the right places (never worked on a complex elisp project before), so please take a closer look at this!

P.S: the approach I have taken is to "dockerize" the existing debug providers and templates (like in `lsp-docker`) so it is easier to understand the code if you are already familiar with that project. Also please take a look at the readme.

### CODE REVIEW

**Comment by factyy** [comment]
Also it has been tested using the Ruby debugger, I can (and will) test a few other debuggers, but (obviously) not all that are listed as supported

**Comment by yyoncho** [comment]
Thank you for the PR. Feel free to add you to the readme as implementor of a major feature. I will try to review it as soon as possible.

File: dap-docker.el [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+(require 'dash)
+(require 'yaml)
+
+(defvar dap-docker-supported-server-types-subtypes
+  (ht ('docker (list 'container 'image)))
+  "A list of all supported server types and subtypes, currently only docker is supported")
+
+(defun dap-docker--is-enabled? (config)
+  "Check whether debugging is enabled"
+  (-if-let* ((server-info (gethash 'debug config))
```
Comment (yyoncho) [inline dap-docker.el:44]: nit: convert to -when-let

File: dap-docker.el [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+             (server-enabled (gethash 'enabled server-info)))
+      (if (equal server-enabled :false)
```
Comment (yyoncho) [inline dap-docker.el:46]: nit:

```
(if (equal server-enabled :false)
          nil
        't)
``` 

is equivalent to `(not (equal server-enabled :false))`

File: dap-docker.el [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            (-any? (lambda (mapping)
+                     (and (f-same? source (car mapping))
+                          (equal destination (cdr mapping))))
+                   (dap-docker--get-path-mappings-from-container container-name)))
+          path-mappings))
+
+(defun dap-docker--get-path-mappings-from-container (container-name)
+  "Get path mappings from a container"
+  (-let ((
+          (inspection-command-program . inspection-command-arguments)
```
Comment (yyoncho) [inline dap-docker.el:142]: nit: on the same line(here and through the review)

File: dap-mode.el [hunk 4]
```diff
@@ -486,19 +490,21 @@ FILE-NAME is the filename in which the breakpoints have been udpated."
           (run-hooks 'dap-breakpoints-changed-hook))))
 
     ;; Update all of the active sessions with the list of breakpoints.
-    (let ((set-breakpoints-req (dap--set-breakpoints-request
-                                file-name
-                                updated-file-breakpoints)))
-      (mapc (lambda (session)
+    (mapc (lambda (session)
+            (debug)
```
Comment (yyoncho) [inline dap-mode.el:494]: leftover

**Review by yyoncho — COMMENTED** [review]
Looks good to me. Several nits and a few questions:

1. This code most likely won't work with adapters that are using tcp, right? If yes, please indicate that in the docs.
2. Can you create a step-by-step hello world example of using the docker stuff? In the spirit of https://emacs-lsp.github.io/lsp-mode/tutorials/CPP-guide/ - you start with Emacs + Docker and you end up with a debug process for debug adapter(feel free to pick the one you use).

File: dap-mode.el (hunk 4 above)
Comment (factyy) [reply dap-mode.el]: Fix in 00c14f8

File: dap-docker.el (hunk 3 above)
Comment (factyy) [reply dap-docker.el]: Fixed in f69dbb1

File: dap-docker.el (hunk 2 above)
Comment (factyy) [reply dap-docker.el]: Fixed in e820d9b

**Comment by factyy** [comment]
@yyoncho , I am a bit concerned about failing tests. Did I forget something?

**Comment by factyy** [comment]
And regarding docs: where do I have to put them?

**Comment by yyoncho** [comment]
> @yyoncho , I am a bit concerned about failing tests. Did I forget something?

Test are red in master.

**Comment by yyoncho** [comment]
> And regarding docs: where do I have to put them?

if you have a blog - it is fine to publish it there. If you don't have a blog you may add entry similar to CPP/Clojure/React tutorials in lsp project.

**Comment by factyy** [comment]
Oh, yeah, missed the branch name. Sorry. Still a bit scary though :)

**Comment by factyy** [comment]
@yyoncho , sorry, couldn't spare any time for writing the docs, will complete it in a few days

**Comment by yyoncho** [comment]
Merging as it is - let's iterate based on the feedback.

**Comment by hmuniz** [comment]
The merge introduced a bug, probably  #641 solves that.

**Pull request merged**